Adding env passthrough

heschlie opened the pull request:
Adding the ability to pass env vars through `calicoctl node run` that are prefixed with `CALICO_*` or FELIX_*`.

Fixes #1501
Fixes #1597

heschlie commented:
@lwr20 Do you think we need more interesting tests for this?  The one I made sets some dummy env vars and verifies they make it into the container but I'm having trouble thinking of a clever way to stress this more if we need it.

tmjd commented:
I think a good test would be to test that the flags take precedence over env vars.

tmjd commented:
If you moved the loop for transferring the env vars to be the first thing put into the envs map then you wouldn't need the check if they already exist.

heschlie commented on tests/st/calicoctl/test_node_run.py:
```diff
@@ -0,0 +1,44 @@
+            felix_out = host.execute('printenv FELIX_TESTVAL')
+
+            self.assertEquals(calico_out, 'dummyVal')
+            self.assertEquals(felix_out, 'felixDummyVal')
```
You are right I do want to be checking the calico-node, I had thought that the DockerHost was the Node container.  Thanks for pointing that out.

lwr20 replied:
You can tell the Dockerhost to start calico-node inside it by simply passing it the argument `start_calico=True` when you create it.

djosborne commented:
Fixes #1501

heschlie commented:
Looking over this in more detail now, the `CALICO_*` will never make it through if we want to prioritize CLI flags (which is definitely what we want) as we default the flags to some value when they are not set.  For example `calicoctl node run --backend=` defaults to `bird` so it always has a value that later gets assigned in the envs map that is passed to the docker command.

Could we instead, not have those set to a default, check for a blank string, then check for an env var, then setting a default if neither are set?  WDPT?  @gunjan5 @lwr20 @tmjd

tmjd commented:
That will work fine for the args that set a string, the boolean ones you'll have to figure something else out because the argutils.ArgBoolOrFalse will always be set in its current implimentation.  Unless there is a way to check if the arg was defaulted or not.

So that is :+1: to what you are proposing but I think there is a little more to switching things around.

heschlie commented:
Sorting out what's causing the tests to fail on Semaphore but not locally...

tomdee commented:
@heschlie What's happening with this PR?

heschlie commented:
@tomdee Sorry for the late reply, finally getting back to this in the last couple days, sorting out why the tests I wrote pass in `make test-containerized` and `make node-test-containerized` but not in `make semaphore`

heschlie commented:
@lwr20 Finally got the tests passing, could I get a review and LGTM if you think it's ready?  I'll squash after that.

robbrockbank commented on tests/st/calicoctl/test_node_run.py:
```diff
@@ -0,0 +1,78 @@
+        Set some CALICO_ and FELIX_ env vars and see that they make it through to the
+        Node container
+        """
+        os.environ['CALICO_TESTVAL'] = 'dummyVal'
```
Rather than setting and unsetting envs, it might be nicer to change the signature of DockerHost calicoctl() method to accept a list of environments ["X=Y", "A=B"] and then prepend the command with exported envs (such as we already do for the ETCD parms)

robbrockbank replied:
I think the approach of pre-pending the command with inline envs also means that we don't risk having dangling environments if the test bails out halfway through.

neiljerram commented:
Just commenting here so that I will be notified when this is merged.  Looking forward to it!

robbrockbank commented:
@heschlie  LGTM, but looks like there is a merge conflict - can you resolve please and I think we should be able to get this checked in.

heschlie commented:
@robbrockbank @neiljerram I've added arguments to the chain of commands down to executing `calicoctl` take a peek and let know WDYT.  

I've fixed the conflicts locally just running the tests again before I push it up.

heschlie commented:
Conflicts removed, tests passed locally, and sqaushed.  Should be ready for merge once Semaphore is green.

robbrockbank commented:
Hmmm, actually I do have a follow up question....

There are a bunch of defaulted values [default: blah] - which all basically map through to environment variables.  Why have we only changed a couple of them to change the inheritance order?

tomdee commented on calicoctl/commands/node/run.go:
```diff
@@ -262,6 +262,17 @@ Description:
+		if strings.HasPrefix(env, "CALICO_") || strings.HasPrefix(env, "FELIX_") {
+			split_env := strings.Split(env, "=")
+			// Let's avoid clobbering anything already set
+			if _, ok := envs[split_env[0]]; !ok {
```
@heschlie Did this get resolved? It looks like maybe it did?

heschlie replied:
It did in the else clause

tomdee commented on calicoctl/commands/node/run.go:
```diff
@@ -205,6 +203,21 @@ Description:
 		// the AS number, so convert.
 		asNumber = argutils.ValidateASNumber(asNumber).String()
 	}
+	// CLI flags will take precedence over env vars
```
I don't understand this change, what is it for?

robbrockbank replied:
It is to re-instate the defaults that were removed from the doc-string.

I think this is because docopt does not distinguish between a value that was explicitly set on the CLI and one that was set through the [default: XXX] option.

robbrockbank replied:
I think the auto-detection methods should also be there now that I think about it.

heschlie replied:
@robbrockbank is correct I had to skip setting a default in the docstring so we could...

1. see if the user set a CLI flag
2. Check if there is an env var if no flag
3. Set a default if neither are present

heschlie replied:
Rob, I can get those added in, I'm going to run through the reference list again as well to see if I missed anything.

heschlie commented:
@robbrockbank @tomdee This is ready for another review, I went ahead and pulled in the upstream changes to my PR and have updated based on your feedback.

neiljerram commented:
@heschlie @robbrockbank @tomdee Is this PR still planned to land sometime soon?  I'm interested so that I can finish off #1587 , which Rob has suggested should be based on this one.  Thanks!

heschlie commented:
@tomdee @robbrockbank I'm going to rebase this again today to get it caught up, could we get this reviewed and merged if possible?

caseydavenport commented:
I'm going to close this since it's pretty stale and I don't think it's a priority right now.

We can re-open/visit this in the future if the requirement arises.
